girder/models/upload.py

```diff
@@ -286,8 +299,9 @@ def createUpload(self, user, name, parentType, parent, size, mimeType=None,
         :type name: str
         :param parentType: The type of the parent being uploaded into.
         :type parentType: str ('folder' or 'item')
-        :param parent: The document representing the parent.
-        :type parentId: dict
+        :param parent: The document representing the parent or the id of the
```

Comment:
Personally, I don't think we should go down this road of model methods accepting an `ObjectId` (or even just a `str`-type ID) in place of a model document. If we do so, we should be consistent about allowing this all across the model API, and such a change would add a huge amount of additional logic to handle all the coercions. As I understand the current design, the model API does not do heavy validation of its parameters and generally takes full documents as parameters.

Furthermore, the fact that the old code was doing `ObjectId(parent['_id'])` seems nonsensical, as a model document should never contain a `str`-typed `_id` field. Indeed, the only time an ID should be `str` typed is right after it's extracted as an API parameter; at that point, it should be passed to `load`, to validate it and do access checking.

So, I think this method should require that `parent` be a `dict`-typed model document. Of course, as an optimization, callers can always pass `parent={'_id': parentId}` when calling the function, if they don't want to do a full lookup of the parent document.